docs/contents/ions.rst

```diff
@@ -107,6 +107,47 @@ This optimization is implemented in NMODL. It can be activated on the CLI via
    nmodl ... codegen --opt-ionvar-copy
 
 
+Concentrations
+--------------
+
+Concentrations adjacent to the inside and outside of the membrane are computed.
+MOD files can change the value by setting ``nai`` or ``nao``. It's considered
+an error if two MOD files write the same concentration, which result in a
+warning.
+
+The reversal potential ``ena`` is a function of the two concentrations.
+Therefore, when writing to either concentration, we must specify if ``ena``
+needs to be recomputed.
+
+Writing Concentrations
+~~~~~~~~~~~~~~~~~~~~~~
+When writing concentrations we must notify NEURON of two things: a) the fact
+that this mechanism writes some concentration; b) that this instance of the
+mechanism does so.
+
+The former happens when registering the mechanism and controls the position of
+the mechanism relative to others, i.e. mechanisms that write the concentration
+must appear before mechanisms that only read the concentration, that way they
+get to see the updated values. We must call ``nrn_writes_conc`` to inform
+NEURON that this mechanisms write some concentration.
+
+The latter notification happens as part of ``nrn_alloc``, i.e. when allocating
+the ``Prop``, by calling ``nrn_check_conc_write``. This checks that there's no
+conflict and modifies the "style" of the ion. We pass ``0`` for exterior and
+``1`` for interior.
+
+Ion Styles
+~~~~~~~~~~
+The style is a 12-bit wide bitfield and stored in the ``dparam`` array of the
+``Prop`` as an ``int``.  It's stored at location ``0`` (variable name:
+``iontype_index_dparam``) in the ``dparam`` array.
+
+From MOD files we need to notify NEURON if we're reading/writing concentration
+or reversal potential. This is done by calling ``nrn_promote(., conc, rev)``.
+The magic value for ``rev`` and ``conc`` is ``1`` for reading and ``3`` for
+writing.
```

Comment:
All of this is a bit unclear for me and I don't know if we could add some small snippet here and there.
If this is irrelevant only ignore this comment.

Reply:
Thank you, I read it again and fixed a mistake and spotted an omission. The difficulty with this text is that in a way this is maybe not the place to document this at all. For example the exact meaning of each of the 12 bits is information internal to NEURON and when generating code we don't need to know, but we do kinda need to know that it's an int and used as a bit-field. Similarly, we're repeating the magic values.

What we do need to know is four things:
1. We must tell NEURON that we'll write to concentrations. Via `nrn_writes_conc` when registering the mech.
2. We must tell NEURON more accurately which concentration we're writing, so it can warn the user if multiple MOD files set the same concentration. Via `nrn_check_conc_write` when allocating the prop.
3. We must inform NEURON (yet again) that we're reading/writing a concentration and/or the reversal potential. Via `nrn_promote` also while allocating the prop.
4. We must tell NEURON (in the INITIAL block only) what we've written the concentration. So it can decide if it needs to recompute the reversal potential. `nrn_wrote_conc`.

In addition there's a few pieces of information that are needed to implement the above requirements. We need to know where the `style` is stored in the ion `Prop` (i.e. at `0`, yes it's a hard-coded constant). I can't help but feel it would be nice to know that exterior is `0` and interior is `1`. Similarly reading is `1` and writing is `3`.